File: packages/soulseek_protocol/lib/src/services/user_service.dart
import 'dart:async';

import '../connection/server_connection.dart';
import '../messages/buffer.dart';
import '../messages/codes.dart';
import '../messages/message.dart';

class UserService {
  final ServerTransport _server;
  final _userStatusController = StreamController<UserStatus>.broadcast();

  late final StreamSubscription _messageSub;

  UserService({required ServerTransport server}) : _server = server;

  Stream<UserStatus> get userStatus => _userStatusController.stream;

  void init() {
    _messageSub = _server.messages.listen(_onMessage);
  }

  void addUser(String username) {
    _server.sendMessage(SoulseekMessage(
      ServerCode.addUser,
      AddUser(username).serialize().toBytes(),
    ));
  }

  void getPeerAddress(String username) {
    _server.sendMessage(SoulseekMessage(
      ServerCode.getPeerAddress,
      GetPeerAddress(username).serialize().toBytes(),
    ));
  }

  void _onMessage(SoulseekMessage message) {
    switch (message.code) {
      case ServerCode.statusResponse:
        _handleStatusResponse(message);
        break;
    }
  }

  void _handleStatusResponse(SoulseekMessage message) {
    try {
      final buffer = ReadBuffer(message.payload);
      final status = UserStatus.parse(buffer);
      _userStatusController.add(status);
    } catch (_) {
      // Skip malformed status
    }
  }

  void dispose() {
    _messageSub.cancel();
    _userStatusController.close();
  }
}

File: Soulseek.Protocol/Services/UserService.cs
using Soulseek.Protocol.Connection;
using Soulseek.Protocol.Messages;

namespace Soulseek.Protocol.Services;

public class UserService
{
    private readonly IServerTransport _server;
    private readonly Subject<UserStatusMessage> _userStatusSubject = new();
    private readonly Subject<PeerAddressResponse> _peerAddressSubject = new();
    private IDisposable? _messageSub;

    public UserService(IServerTransport server)
    {
        _server = server;
    }

    public IObservable<UserStatusMessage> UserStatus => _userStatusSubject;
    public IObservable<PeerAddressResponse> PeerAddressResponses => _peerAddressSubject;

    public void Init()
    {
        _messageSub = _server.MessageStream.Subscribe(OnMessage);
    }

    public void AddUser(string username)
    {
        _server.SendMessage(new SoulseekMessage(
            ServerCode.AddUser,
            new AddUser(username).Serialize().ToBytes()
        ));
    }

    public void GetPeerAddress(string username)
    {
        _server.SendMessage(new SoulseekMessage(
            ServerCode.GetPeerAddress,
            new GetPeerAddress(username).Serialize().ToBytes()
        ));
    }

    private void OnMessage(SoulseekMessage message)
    {
        try
        {
            if (message.Code == ServerCode.StatusResponse)
            {
                var buffer = new ReadBuffer(message.Payload);
                var status = UserStatusMessage.Parse(buffer);
                _userStatusSubject.OnNext(status);
            }
            else if (message.Code == ServerCode.GetPeerAddress)
            {
                var buffer = new ReadBuffer(message.Payload);
                var addr = PeerAddressResponse.Parse(buffer);
                _peerAddressSubject.OnNext(addr);
            }
        }
        catch
        {
            // Skip malformed messages
        }
    }

    public void Dispose()
    {
        _messageSub?.Dispose();
    }
}